File: momentra/momentra/Shell/GroupActive/Experience/Create/ExperienceQuickAddSheets.swift
import SwiftUI

/// Figma party 592:8580 / outing 592:7770 — moment-colored Experience Quick Add sheets.
struct ExperienceGapQuickAddSheet: View {
    let theme: ExperienceActiveTheme
    let kind: ExperienceQuickAddKind
    var momentId: String? = nil
    var momentTypeCode: String? = nil
    var onClose: () -> Void
    var onSaved: () -> Void = {}
    var onBooking: () -> Void = {}

    private var accent: SheetAccent {
        SheetAccent(accent: theme.accent, accentEnd: theme.accentSolid, soft: theme.accentSoft)
    }

    var body: some View {
        if kind == .booking {
            Color.clear
                .onAppear {
                    onBooking()
                    onClose()
                }
        } else if kind == .expense, let momentId {
            GroupExpenseSheet(
                momentId: momentId,
                isPresented: Binding(
                    get: { true },
                    set: { if !$0 { onClose() } }
                ),
                momentTypeCode: momentTypeCode,
                onSaved: {
                    onSaved()
                    onClose()
                }
            )
        } else {
            NativeSheetScaffold(
                title: kind.label,
                onClose: onClose,
                background: Color(hex: "#1C1A24")
            ) {
                ScrollView {
                    sheetBody
                        .padding(.horizontal, 20)
                        .padding(.top, 16)
                }
            }
            .presentationDetents([.large])
            .presentationDragIndicator(.visible)
        }
    }

    @ViewBuilder
    private var sheetBody: some View {
        switch kind {
        case .expense:
            EmptyView()
        case .contribution:
            WeddingContributionBody(momentId: momentId, onDismiss: onClose, onSaved: onSaved, accent: accent)
        case .budget:
            WeddingBudgetBody(momentId: momentId, onDismiss: onClose, onSaved: onSaved, accent: accent)
        case .participant:
            ExperienceParticipantBody(momentId: momentId, theme: theme, onDismiss: onClose, onSaved: onSaved, accent: accent)
        case .vendor:
            WeddingVendorBody(momentId: momentId, onDismiss: onClose, onSaved: onSaved, accent: accent)
        case .planning:
            WeddingPlanningBody(momentId: momentId, momentTypeCode: momentTypeCode, onDismiss: onClose, onSaved: onSaved, accent: accent)
        case .attendance:
            WeddingAttendanceBody(momentId: momentId, onDismiss: onClose, onSaved: onSaved, accent: accent)
        case .poll:
            WeddingPollBody(momentId: momentId, onDismiss: onClose, onSaved: onSaved, accent: accent)
        case .memory:
            WeddingMemoryBody(momentId: momentId, onDismiss: onClose, onSaved: onSaved, accent: accent)
        case .update:
            WeddingUpdateBody(momentId: momentId, onDismiss: onClose, onSaved: onSaved, accent: accent)
        case .settle:
            WeddingSettleBody(momentId: momentId, onDismiss: onClose, onSaved: onSaved)
        case .booking:
            EmptyView()
        }
    }
}

private struct ExperienceParticipantBody: View {
    var momentId: String?
    let theme: ExperienceActiveTheme
    var onDismiss: () -> Void
    var onSaved: () -> Void
    var accent: SheetAccent

    @State private var name = ""
    @State private var contact = ""
    @State private var role = ""
    @State private var rsvp = "Pending"
    @State private var plusOne = false
    @State private var notes = ""
    @State private var busy = false
    @State private var error: String?

    private var canSave: Bool {
        momentId != nil && !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && !busy
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            SheetHeader(
                icon: "person.2.fill",
                title: "Add Participant",
                subtitle: theme.participantSubtitle,
                accent: accent
            )

            VStack(alignment: .leading, spacing: 8) {
                FieldLabel(text: "Participant Name")
                SheetField(value: $name, placeholder: "Full name", minHeight: 42)
            }

            VStack(alignment: .leading, spacing: 8) {
                FieldLabel(text: "Email or Phone")
                SheetField(value: $contact, placeholder: "Email or phone", minHeight: 42, keyboardType: .emailAddress)
            }

            VStack(alignment: .leading, spacing: 8) {
                FieldLabel(text: "Role")
                ChipRow(options: theme.participantRoles, selected: $role, accent: accent)
            }

            VStack(alignment: .leading, spacing: 8) {
                FieldLabel(text: "RSVP Status")
                ChipRow(options: ["Confirmed", "Pending", "Declined"], selected: $rsvp, accent: accent)
            }

            HStack {
                VStack(alignment: .leading, spacing: 2) {
                    Text("Plus One Allowed")
                        .font(.plusJakarta(size: 14, weight: .semibold))
                        .foregroundStyle(Color(hex: "#FFFFFF"))
                    Text("Include guest's spouse or partner")
                        .font(.plusJakarta(size: 12))
                        .foregroundStyle(Color(hex: "#9E9AA8"))
                }
                Spacer()
                Toggle("", isOn: $plusOne)
                    .labelsHidden()
                    .tint(accent.accent)
            }
            .padding(.vertical, 4)

            VStack(alignment: .leading, spacing: 8) {
                FieldLabel(text: "Dietary Preferences / Notes")
                SheetField(value: $notes, placeholder: "Optional notes", minHeight: 42)
            }

            if let error {
                Text(error)
                    .font(.plusJakarta(size: 12))
                    .foregroundStyle(Color(hex: "#F87171"))
            }

            PrimaryCta(
                label: busy ? "Saving…" : "Add Participant",
                enabled: canSave,
                accent: accent,
                loading: busy,
                lightLabel: true
            ) {
                Task { await save() }
            }
        }
        .onAppear {
            if role.isEmpty {
                role = theme.participantRoles.first ?? "Guest"
            }
        }
    }

    private func save() async {
        guard let momentId else { return }
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        busy = true
        error = nil
        let contactTrim = contact.trimmingCharacters(in: .whitespacesAndNewlines)
        let email = contactTrim.contains("@") ? contactTrim : nil
        let phone = email == nil && !contactTrim.isEmpty ? contactTrim : nil
        do {
            _ = try await APIClient.shared.addGroupParticipant(
                momentId: momentId,
                displayName: trimmed,
                roleCode: "PARTICIPANT",
                email: email,
                phone: phone
            )
            onSaved()
            onDismiss()
        } catch {
            self.error = error.localizedDescription
        }
        busy = false
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/group/experience/create/ExperienceQuickAddSheets.kt
package com.example.momentra.ui.shell.group.experience.create

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.momentra.R
import com.example.momentra.data.repository.GroupSliceRepository
import com.example.momentra.ui.shell.group.shared.GroupExpenseSheet
import com.example.momentra.ui.shell.group.shared.TripSheetTokens
import com.example.momentra.ui.shell.group.wedding.create.ChipRow
import com.example.momentra.ui.shell.group.wedding.create.FieldLabel
import com.example.momentra.ui.shell.group.wedding.create.PrimaryCta
import com.example.momentra.ui.shell.group.wedding.create.SheetAccent
import com.example.momentra.ui.shell.group.wedding.create.SheetField
import com.example.momentra.ui.shell.group.wedding.create.SheetHeader
import com.example.momentra.ui.shell.group.wedding.create.WeddingAttendanceSheetBody
import com.example.momentra.ui.shell.group.wedding.create.WeddingBudgetSheetBody
import com.example.momentra.ui.shell.group.wedding.create.WeddingContributionSheetBody
import com.example.momentra.ui.shell.group.wedding.create.WeddingMemorySheetBody
import com.example.momentra.ui.theme.PlusJakartaSans
import kotlinx.coroutines.launch
import com.example.momentra.ui.shell.group.wedding.create.WeddingPlanningSheetBody
import com.example.momentra.ui.shell.group.wedding.create.WeddingPollSheetBody
import com.example.momentra.ui.shell.group.wedding.create.WeddingSettleSheetBody
import com.example.momentra.ui.shell.group.wedding.create.WeddingUpdateSheetBody
import com.example.momentra.ui.shell.group.wedding.create.WeddingVendorSheetBody

private fun ExperienceActiveTheme.toSheetAccent(): SheetAccent =
    SheetAccent(accent = TripSheetTokens.Accent, accentEnd = TripSheetTokens.AccentEnd, soft = accentSoft)

/** Figma 575:15497 — moment-colored Experience Quick Add sheets. */
@OptIn(ExperimentalMaterial3Api::class, ExperimentalLayoutApi::class)
@Composable
fun ExperienceGapQuickAddSheet(
    theme: ExperienceActiveTheme,
    kind: ExperienceQuickAddKind,
    visible: Boolean,
    momentId: String?,
    onDismiss: () -> Unit,
    onSaved: () -> Unit = {},
    onBooking: () -> Unit = {},
    momentTypeCode: String? = null,
    repository: GroupSliceRepository = remember { GroupSliceRepository() },
) {
    if (!visible) return
    if (kind == ExperienceQuickAddKind.BOOKING) {
        LaunchedEffect(kind) {
            onBooking()
            onDismiss()
        }
        return
    }
    if (kind == ExperienceQuickAddKind.EXPENSE && !momentId.isNullOrBlank()) {
        GroupExpenseSheet(
            momentId = momentId,
            visible = true,
            onDismiss = onDismiss,
            onSaved = {
                onSaved()
                onDismiss()
            },
            momentTypeCode = momentTypeCode,
            repository = repository,
        )
        return
    }

    val accent = theme.toSheetAccent()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = TripSheetTokens.Bg,
        dragHandle = {
            Box(
                modifier = Modifier
                    .padding(top = 12.dp, bottom = 4.dp)
                    .size(width = 40.dp, height = 5.dp)
                    .clip(RoundedCornerShape(100.dp))
                    .background(TripSheetTokens.Border),
            )
        },
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .navigationBarsPadding()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 20.dp)
                .padding(top = 4.dp, bottom = 28.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            when (kind) {
                ExperienceQuickAddKind.EXPENSE -> Unit
                ExperienceQuickAddKind.CONTRIBUTION ->
                    WeddingContributionSheetBody(momentId, repository, onDismiss, onSaved, accent)
                ExperienceQuickAddKind.BUDGET ->
                    WeddingBudgetSheetBody(momentId, repository, onDismiss, onSaved, accent)
                ExperienceQuickAddKind.PARTICIPANT ->
                    ExperienceParticipantSheetBody(momentId, repository, onDismiss, onSaved, theme, accent)
                ExperienceQuickAddKind.VENDOR ->
                    WeddingVendorSheetBody(momentId, repository, onDismiss, onSaved, accent)
                ExperienceQuickAddKind.PLANNING ->
                    WeddingPlanningSheetBody(
                        momentId,
                        repository,
                        onDismiss,
                        onSaved,
                        accent,
                        momentTypeCode = momentTypeCode,
                    )
                ExperienceQuickAddKind.ATTENDANCE ->
                    WeddingAttendanceSheetBody(momentId, repository, onDismiss, onSaved, accent)
                ExperienceQuickAddKind.POLL ->
                    WeddingPollSheetBody(momentId, repository, onDismiss, onSaved, accent)
                ExperienceQuickAddKind.MEMORY ->
                    WeddingMemorySheetBody(momentId, repository, onDismiss, onSaved, accent)
                ExperienceQuickAddKind.UPDATE ->
                    WeddingUpdateSheetBody(momentId, repository, onDismiss, onSaved, accent)
                ExperienceQuickAddKind.SETTLE ->
                    WeddingSettleSheetBody(momentId, onDismiss, onSaved)
                ExperienceQuickAddKind.BOOKING -> Unit
            }
        }
    }
}

@Composable
private fun ExperienceParticipantSheetBody(
    momentId: String?,
    repository: GroupSliceRepository,
    onDismiss: () -> Unit,
    onSaved: () -> Unit,
    theme: ExperienceActiveTheme,
    accent: SheetAccent,
) {
    var name by remember { mutableStateOf("") }
    var contact by remember { mutableStateOf("") }
    var role by remember { mutableStateOf(theme.participantRoles.first()) }
    var rsvp by remember { mutableStateOf("Pending") }
    var plusOne by remember { mutableStateOf(false) }
    var notes by remember { mutableStateOf("") }
    var submitting by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()

    SheetHeader(
        R.drawable.ic_qa_users,
        "Add Participant",
        theme.participantSubtitle,
        accent = accent,
    )
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        FieldLabel("Participant Name")
        SheetField(name, { name = it }, "Full name", minHeight = 42)
    }
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        FieldLabel("Email or Phone")
        SheetField(contact, { contact = it }, "Email or phone", minHeight = 42)
    }
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        FieldLabel("Role")
        ChipRow(theme.participantRoles, role, accent) { role = it }
    }
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        FieldLabel("RSVP Status")
        ChipRow(listOf("Confirmed", "Pending", "Declined"), rsvp, accent) { rsvp = it }
    }
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        FieldLabel("Dietary Preferences / Notes")
        SheetField(notes, { notes = it }, "Optional notes", minHeight = 42)
    }
    error?.let { Text(it, color = Color(0xFFF87171), fontSize = 12.sp, fontFamily = PlusJakartaSans) }
    PrimaryCta(
        label = if (submitting) "Saving…" else "Add Participant",
        enabled = !momentId.isNullOrBlank() && name.isNotBlank() && !submitting,
        loading = submitting,
        accent = accent,
        lightLabel = true,
        onClick = {
            val id = momentId ?: return@PrimaryCta
            scope.launch {
                submitting = true
                error = null
                val trimmedContact = contact.trim()
                val email = if (trimmedContact.contains("@")) trimmedContact else null
                val phone = if (email == null && trimmedContact.isNotBlank()) trimmedContact else null
                repository.addParticipant(
                    id,
                    name.trim(),
                    roleCode = "PARTICIPANT",
                    email = email,
                    phone = phone,
                ).fold(
                    onSuccess = {
                        submitting = false
                        onSaved()
                        onDismiss()
                    },
                    onFailure = {
                        submitting = false
                        error = it.message
                    },
                )
            }
        },
    )
}
